io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -797,7 +919,58 @@ unblockThreads !wakeup !simstate@SimState {runqueue, threads} =
     -- and in which case we mark them as now running
     !threads'  = List.foldl'
                    (flip (Map.adjust (\t -> t { threadBlocked = False })))
-                  threads unblocked
+                   threads
+                   unblocked
+
+
+throwExceptionThreads :: [(ThreadId, TimeoutId)] -> SimState s a -> SimState s a
+throwExceptionThreads timeoutExpired simState@SimState {threads, locks} =
+  foldl' (\st@SimState{ runqueue = runqueue
+                      , threads  = threads'
+                      , locks    = locks'
+                      }
+           (t, tmid)
+          -> let tid'      = threadId t
+                 threads'' = Map.insert tid' t threads'
+                 locks''   = Map.update (const (Just (Locked tid'))) tmid locks'
+                 runqueue' = Deque.snoc tid' runqueue
+              in st { runqueue = runqueue'
+                    , threads  = threads''
+                    , locks    = locks''
+                    })
+          simState
+          throwToThread
+
+  where
+    -- can only throw exception if the thread exists and if the mutually
+    -- exclusive lock exists and is still 'NotLocked'
+    toThrow = [ (tid, tmid, t)
+              | (tid, tmid) <- timeoutExpired
+              , Just t <- [Map.lookup tid threads]
+              , Just NotLocked <- [Map.lookup tmid locks]
+              ]
+    -- we launch a thread responsible for throwing an AsyncCancelled exception
+    -- to the thread which timeout expired
+    throwToThread =
+      [ let nextId   = threadNextTId t
+            tid'     = childThreadId tid nextId
+         in ( Thread { threadId      = tid'
+                    , threadControl =
+                       ThreadControl
+                         (ThrowTo (toException (TimeoutException tmid))
+                                   tid
+                                   (Return ()))
+                         ForkFrame
+                    , threadBlocked = False
+                    , threadMasking = threadMasking t
```

Comment:
I think the new thread _must_ start with async exceptions unmasked, rather than inheriting the masking state from the parent thread. This is important because the current locking scheme relies on being able to kill this thread, which is only possible if async exceptions are unmasked (or at least not masked uninterruptible).

See the GHC timeout code that uses `forkIOWithUnmask $ \unmask -> unmask $ do`

Reply:
Yes, good catch!